lib/ansible/template/native_helpers.py

```diff
@@ -28,6 +34,33 @@ def visit_Name(self, node):
         return ast.Constant(value=_JSON_MAP[node.id])
 
 
+def _is_unsafe(value: t.Any) -> bool:
+    """
+    Our helper function, which will also recursively check dict and
+    list entries due to the fact that they may be repr'd and contain
+    a key or value which contains jinja2 syntax and would otherwise
+    lose the AnsibleUnsafe value.
+    """
+    to_check = [value]
+
+    while True:
+        if not to_check:
```

Comment:
Possible to hit an object cycle here, so might need to accumulate a set of seen object identities?

Reply:
I can certainly add the set as a check, this shouldn't be any problem AFAIK as the original recursion code didn't do any id checks.